optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsConstraintStreamScoreDirectorFactory.java

```diff
@@ -151,15 +150,21 @@ private static KieBase buildKieBaseFromModel(Model model, boolean droolsAlphaNet
         ScoreDefinition<Score_> scoreDefinition = solutionDescriptor.getScoreDefinition();
         ScoreInliner<Score_> scoreInliner =
                 scoreDefinition.buildScoreInliner((Map) constraintToWeightMap, constraintMatchEnabled);
-        constraintToWeightMap.forEach((constraint, weight) -> {
-            if (Objects.equals(weight, zeroScore)) {
-                return;
+        Set<String> disabledConstraints = new HashSet<>();
+        for (Map.Entry<DroolsConstraint<Solution_>, Score_> entry : constraintToWeightMap.entrySet()) {
+            DroolsConstraint<Solution_> constraint = entry.getKey();
+            Score_ constraintWeight = entry.getValue();
+            if (constraintWeight.isZero()) {
+                disabledConstraints.add(constraint.getConstraintId());
+            } else {
+                String globalName = kieBaseDescriptor.getConstraintToGlobalMap().get(constraint).getName();
+                kieSession.setGlobal(globalName, scoreInliner.buildWeightedScoreImpacter(constraint));
             }
-            String globalName = kieBaseDescriptor.getConstraintToGlobalMap().get(constraint).getName();
-            kieSession.setGlobal(globalName, scoreInliner.buildWeightedScoreImpacter(constraint));
-        });
+        }
         // Return only the inliner as that holds the work product of the individual impacters.
-        return new SessionDescriptor<>(kieSession, new ConstraintDisablingAgendaFilter(constraintToWeightMap), scoreInliner);
+        AgendaFilter agendaFilter =
+                disabledConstraints.isEmpty() ? null : new ConstraintDisablingAgendaFilter(Set.copyOf(disabledConstraints));
```

Comment:
The agenda filter execution will be skipped entirely when there are no rules to disable.